{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import absolute_import\n",
    "from __future__ import division\n",
    "from __future__ import print_function\n",
    "\n",
    "import sys\n",
    "import itertools as it\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam\n",
    "\n",
    "\n",
    "class Genotypes(object):\n",
    "    \"\"\"A class for storing genotype data which at its core \n",
    "    consists of a dense p x n genotype matrix, a dataframe \n",
    "    of metadata associated which each snp, and a list of \n",
    "    individual ids. We also provide computation options for \n",
    "    setting up dataset for commonly using analyses\n",
    "    \n",
    "    Arguments\n",
    "    ---------\n",
    "    Y : np.array\n",
    "        p x n genotype matrix \n",
    "        \n",
    "    snp_df : pd.DataFrame\n",
    "        Dataframe storing snp level meta data\n",
    "        \n",
    "    inds : list\n",
    "        list of individual ids\n",
    "    \n",
    "    Attributes\n",
    "    ----------\n",
    "    Y : np.array\n",
    "        p x n genotype matrix \n",
    "        \n",
    "    snp_df : pd.DataFrame\n",
    "        Dataframe storing snp level meta data\n",
    "        \n",
    "    inds : list\n",
    "        list of individual ids\n",
    "        \n",
    "    n : int\n",
    "        number of individuals (samples)\n",
    "        \n",
    "    p : int \n",
    "        number of snps (features)\n",
    "        \n",
    "    base_complement : dict\n",
    "        dictionary where keys are bases and values\n",
    "        are the bases complement\n",
    "        \n",
    "    bases : list\n",
    "        list of bases\n",
    "        \n",
    "    strand_ambiguous_alleles : dict\n",
    "        dictionary of strand ambiguous alleles \n",
    "        combinations as keys and bools as values\n",
    "    \"\"\"\n",
    "    def __init__(self, Y=None, snp_df=None, inds=None):\n",
    "\n",
    "        # p x n genotype matrix\n",
    "        self.Y = Y\n",
    "\n",
    "        # DataFrame of SNP level information\n",
    "        self.snp_df = snp_df\n",
    "        \n",
    "        # list of individual ids\n",
    "        self.inds = inds\n",
    "\n",
    "        if ((type(Y) != type(None)) and \n",
    "            (type(snp_df) != type(None)) and \n",
    "            (type(inds) != type(None))):\n",
    "            # number of individuals\n",
    "            self.n = len(self.inds)\n",
    "\n",
    "            # number of snps\n",
    "            self.p = self.snp_df.shape[0]\n",
    "        else:\n",
    "            # number of individuals\n",
    "            self.n = None\n",
    "\n",
    "            # number of snps\n",
    "            self.p = None\n",
    "\n",
    "        # below is adapted from \n",
    "        # https://github.com/bulik/ldsc/blob/master/munge_sumstats.py\n",
    "\n",
    "        # complement of each base\n",
    "        self.base_complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
    "\n",
    "        # bases \n",
    "        self.bases = self.base_complement.keys()\n",
    "\n",
    "        # dict of allele combinations that are strand ambiguous\n",
    "        self.strand_ambiguous_alleles = {''.join(x): x[0] == self.base_complement[x[1]]\n",
    "                                         for x in it.product(self.bases, self.bases)\n",
    "                                         if x[0] != x[1]} \n",
    "\n",
    "    def remove_chrom_prefix(self):\n",
    "        \"\"\"Remove the chromsome prefix from the snp_df\n",
    "        this helps for making datasets consistent before \n",
    "        merging\n",
    "        \"\"\"\n",
    "        self.snp_df[\"CHROM\"] = self.snp_df[\"CHROM\"].apply(lambda x: int(x[3:]))\n",
    "\n",
    "    def keep_autosomes(self):\n",
    "        \"\"\"Keep snps only present on the autosomes\n",
    "        \"\"\"\n",
    "        # filter out snps not on the autosome\n",
    "        self.snp_df = self.snp_df[self.snp_df[\"CHROM\"] <= 22]\n",
    "\n",
    "    def remove_rare_common(self, eps):\n",
    "        \"\"\"Removes SNPs that are too rare or \n",
    "        too common given some threshold eps\n",
    "        Arguments\n",
    "        ---------\n",
    "        eps : float\n",
    "            frequency threshold eps\n",
    "        \"\"\"\n",
    "        # compute allele frequencies at each SNP\n",
    "        f = np.nansum(self.Y, axis=1) / (2 * self.n)\n",
    "\n",
    "        # get the indicies of SNPs that dont pass the thresh\n",
    "        idx = np.where((f >= eps) & (f <= (1. - eps)))[0]\n",
    "        \n",
    "        # keep only these SNPs \n",
    "        self.snp_df = self.snp_df.iloc[idx]\n",
    "    \n",
    "    def remove_strand_ambiguous_alleles(self):\n",
    "        \"\"\"Removes any snps that have alleles that\n",
    "        are strand ambiguous. \n",
    "        TODO: speed this up its currently a bit slow\n",
    "        also add removal of incosistent alleles\n",
    "        \"\"\"\n",
    "        # lambda function for applying to each snp\n",
    "        strand_unamb_func = lambda row: not self.strand_ambiguous_alleles[row[\"A1\"] + row[\"A2\"]]\n",
    "\n",
    "        # snps that have unambigous strand\n",
    "        idx = self.snp_df.apply(strand_unamb_func, axis=1)\n",
    "\n",
    "        # index the the snps\n",
    "        self.snp_df = self.snp_df[idx]\n",
    "    \n",
    "    def reindex(self):\n",
    "        \"\"\"Reset the index column in the pandas dataframe\n",
    "        \"\"\"\n",
    "        # get indicies of snps to keep\n",
    "        idx = self.snp_df[\"idx\"].tolist()\n",
    "\n",
    "        # reindex data matrix\n",
    "        self.Y = self.Y[idx, :] \n",
    "\n",
    "        # re-compute number of snps\n",
    "        self.p = self.snp_df.shape[0]\n",
    "\n",
    "        # reset index column\n",
    "        self.snp_df[\"idx\"] = np.arange(0, self.p, 1)\n",
    "    \n",
    "    def binarize(self):\n",
    "        \"\"\"Convert genotype data to binary\n",
    "        by mapping 2->1, 0->0 and randomly \n",
    "        selecting a 0 or 1 for heterozygotes.\n",
    "        This emulates the read sampled data found\n",
    "        in ancient DNA studies\n",
    "        \"\"\"\n",
    "        # randomly sample hets\n",
    "        p_het = np.sum(self.Y[self.Y == 1.])\n",
    "        if p_het > 0:\n",
    "            samps = np.random.binomial(1, .5, int(p_het)).astype(np.float32) \n",
    "            np.place(self.Y, self.Y == 1., samps)\n",
    "\n",
    "        # set 2s to 1s\n",
    "        self.Y[self.Y == 2.] = 1.\n",
    "\n",
    "    def normalize(self, scale, impute):\n",
    "        \"\"\"Mean center the data so each snp has mean 0. \n",
    "        If scale true then scale each SNP to have var 1. \n",
    "        If impute is true it sets missing sites to 0. i.e. \n",
    "        the approx mean. Note for all computations we \n",
    "        ignore nans.\n",
    "        Arguments\n",
    "        ---------\n",
    "        scale : bool\n",
    "            scale the data to have variance 1\n",
    "        impute : bool\n",
    "            predict missing sites with the expected mean 0\n",
    "        \"\"\"\n",
    "        if scale:\n",
    "            # compute mean genotype for every SNP\n",
    "            mu = np.nanmean(self.Y, axis=1)\n",
    "\n",
    "            # compute std dev for every SNP\n",
    "            std = np.nanstd(self.Y, axis=1)\n",
    "\n",
    "            # center and scale\n",
    "            Zt = (self.Y.T - mu) / std\n",
    "            self.Z = Zt.T\n",
    "        else:\n",
    "            # compute mean genotype for every SNP\n",
    "            mu = np.nanmean(self.Y, axis=1)\n",
    "\n",
    "            # center\n",
    "            Zt = self.Y.T - mu\n",
    "            self.Z = Zt.T\n",
    "\n",
    "        if impute:\n",
    "            # set nans to zero\n",
    "            self.Z = np.nan_to_num(self.Z)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
